File: 2025/day02.py
def gen_repeat(i):
    match i:
        case 2:
            for j in range(10):
                yield int(f"{j}{j}")
        case 3:
            for j in range(10):
                yield int(f"{j}{j}{j}")
        case 4:
            for j in range(10):
                yield int(f"{j}{j}{j}{j}")
            for j in range(10, 100):
                yield int(f"{j}{j}")
        case 5:
            for j in range(10):
                yield int(f"{j}{j}{j}{j}{j}")
        case 6:
            for j in range(10, 100):
                yield int(f"{j}{j}{j}")
            for j in range(100, 1000):
                yield int(f"{j}{j}")
        case 7:
            for j in range(10):
                yield int(f"{j}{j}{j}{j}{j}{j}{j}")
        case 8:
            for j in range(1000, 10000):
                yield int(f"{j}{j}")
        case 9:
            for j in range(100, 1000):
                yield int(f"{j}{j}{j}")
        case 10:
            for j in range(10, 100):
                yield int(f"{j}{j}{j}{j}{j}")
            for j in range(10000, 100000):
                yield int(f"{j}{j}")


with open("2025/input/day02") as f:
    data = f.read().splitlines()[0]
    raw_ranges = data.split(",")
    ranges = [x.split("-") for x in raw_ranges]

score1 = 0
score2 = 0

for r in ranges:
    if len(r[0]) % 2 == 0 or len(r[1]) % 2 == 0 or (len(r[1]) - len(r[0])) >= 2:
        if len(r[0]) % 2:
            start = r[0].rjust(len(r[0]) + 1, "0")
        else:
            start = r[0]
        if len(r[1]) % 2:
            end = r[1].ljust(len(r[1]) + 1, "0")
        else:
            end = r[1]
        start = int(start[0 : len(start) // 2])
        end = int(end[0 : len(end) // 2])

        for i in range(start, end + 1):
            val = int(f"{i}{i}")
            if int(r[0]) <= val <= int(r[1]):
                score1 += val

print(f"Partie 1: {score1}")

for r in ranges:
    for size in range(len(r[0]), len(r[1]) + 1):
        seen = set()
        for val in gen_repeat(size):
            if val not in seen and int(r[0]) <= val <= int(r[1]):
                seen.add(val)
                score2 += val

print(f"Partie 2: {score2}")

File: 2025/rust/src/aoc2025/day02.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::cmp::max;
use std::fs::File;

#[derive(Debug)]
struct IDRange {
    low: u64,
    high: u64,
    invalid_id: u64,
    seen: Vec<u64>,
}

impl IDRange {
    fn new(low: u64, high: u64) -> IDRange {
        IDRange {
            low: low,
            high: high,
            invalid_id: 0,
            seen: Vec::new(),
        }
    }
}

impl Iterator for IDRange {
    type Item = u64;

    fn next(&mut self) -> Option<Self::Item> {
        let nb_digits_low = max(self.low.to_string().len(), 2);
        let nb_digits_high = self.high.to_string().len();

        for nb_digits in nb_digits_low..=nb_digits_high {
            match nb_digits {
                2 => {
                    for i in (11..100).step_by(11) {
                        if i >= self.low && i <= self.high && !self.seen.contains(&i) {
                            self.invalid_id = i;
                            self.seen.push(i);
                            return Some(i);
                        }
                    }
                }
                3 => {
                    for i in (111..1000).step_by(111) {
                        if i >= self.low && i <= self.high && !self.seen.contains(&i) {
                            self.invalid_id = i;
                            self.seen.push(i);
                            return Some(i);
                        }
                    }
                }
                4 => {
                    for i in (1111..10000).step_by(1111) {
                        if i >= self.low && i <= self.high && !self.seen.contains(&i) {
                            self.invalid_id = i;
                            self.seen.push(i);
                            return Some(i);
                        }
                    }
                    for i in 10..100 {
                        let n = i * 100 + i;
                        if n >= self.low && n <= self.high && !self.seen.contains(&n) {
                            self.invalid_id = n;
                            self.seen.push(n);
                            return Some(n);
                        }
                    }
                }
                5 => {
                    for i in (11111..100000).step_by(11111) {
                        if i >= self.low && i <= self.high && !self.seen.contains(&i) {
                            self.invalid_id = i;
                            self.seen.push(i);
                            return Some(i);
                        }
                    }
                }
                6 => {
                    for i in 10..100 {
                        let n = i * 10000 + i * 100 + i;
                        if n >= self.low && n <= self.high && !self.seen.contains(&n) {
                            self.invalid_id = n;
                            self.seen.push(n);
                            return Some(n);
                        }
                    }
                    for i in 100..1000 {
                        let n = i * 1000 + i;
                        if n >= self.low && n <= self.high && !self.seen.contains(&n) {
                            self.invalid_id = n;
                            self.seen.push(n);
                            return Some(n);
                        }
                    }
                }
                7 => {
                    for i in (1111111..10000000).step_by(1111111) {
                        if i >= self.low && i <= self.high && !self.seen.contains(&i) {
                            self.invalid_id = i;
                            self.seen.push(i);
                            return Some(i);
                        }
                    }
                }
                8 => {
                    for i in 1000..10000 {
                        let n = i * 10000 + i;
                        if n >= self.low && n <= self.high && !self.seen.contains(&n) {
                            self.invalid_id = n;
                            self.seen.push(n);
                            return Some(n);
                        }
                    }
                }
                9 => {
                    for i in 100..1000 {
                        let n = i * 1000000 + i * 1000 + i;
                        if n >= self.low && n <= self.high && !self.seen.contains(&n) {
                            self.invalid_id = n;
                            self.seen.push(n);
                            return Some(n);
                        }
                    }
                }
                10 => {
                    for i in 10..100 {
                        let n = i * 100000000 + i * 1000000 + i * 10000 + i * 100 + i;
                        if n >= self.low && n <= self.high && !self.seen.contains(&n) {
                            self.invalid_id = n;
                            self.seen.push(n);
                            return Some(n);
                        }
                    }
                    for i in 10000..100000 {
                        let n = i * 100000 + i;
                        if n >= self.low && n <= self.high && !self.seen.contains(&n) {
                            self.invalid_id = n;
                            self.seen.push(n);
                            return Some(n);
                        }
                    }
                }
                _ => return None,
            }
        }
        return None;
    }
}

fn simple_invalid_id(val: u64) -> bool {
    let id_str = val.to_string();
    if id_str.len() % 2 == 1 {
        return false;
    }
    // check if the first half of the string is equal to the second half
    if id_str[0..id_str.len() / 2] == id_str[id_str.len() / 2..] {
        return true;
    }
    false
}

pub fn day02(path: &String) {
    let input: String = File::open(path).read_lines::<String>(1).next().unwrap();

    let mut data: Vec<IDRange> = input
        .split(",")
        .map(|s| {
            s.split_once('-')
                .map(|(a, b)| IDRange::new(a.parse::<u64>().unwrap(), b.parse::<u64>().unwrap()))
                .unwrap()
        })
        .collect();

    let mut score1 = 0;
    let mut score2 = 0;

    for r in &mut data {
        r.for_each(|i| {
            score2 += i;
            if simple_invalid_id(i) {
                score1 += i;
            }
        })
    }

    println!("Part 1: {}", score1);
    println!("Part 2: {}", score2);
}
